### PR TITLE
[WIP] remove sorting in encoders

### PR DESCRIPTION
**Pull request by DeNeutoy**
`torch.nn.utils.rnn.pack_padded_sequence` now accepts unsorted sequence tensors, which means we no-longer have to sort them upfront, meaning we can remove loads of hideously complicated indexing code 👍 

Changes:

- we no longer sort input tensors by sequence length
- we no longer remove tensors which are completely padded. Instead, we (slightly hackily) set their sequence length to 1, and then mask out the output.
- we no longer have to manually append padding onto the sequence length dimension, because this can now be specified in the `total_length` argument to `pad_packed_sequence`.

### CODE REVIEW

**Comment by matt-gardner** [comment]
Does not sorting have any performance implications?  Presumably the logic on our side is faster, because we don't do as much. What about on the pytorch side?

**Comment by DeNeutoy** [comment]
Good point, I am checking on the pytorch slack now

**Comment by DeNeutoy** [comment]
Turns out this is infeasible because we have several classes which assume the packed sequence is sorted. Damn.
